### PR TITLE
Add VolumeHistogram vector postprocessor

### LINKED ISSUE #7801: Add volume histogram VectorPostprocessor
### Description of the enhancement or error report

For my app I need to determine the amount of volume in the simulation cell that has a variable at a certain value, and I need this data as a function of the value. I.e. I need to create a volume histogram over a variable. This seems general enough to go into one of the modules. Potentially heat conduction (I need temperature statistics).

### Rationale for the enhancement or information for reproducing the error

Added capability.

### Identified impact

Improved postprocessing.

### PR DESCRIPTION
**Pull request by dschwen**
Let's the user obtain a histogram of the volume fractions occupied by a certain variable value.

Closes #7801

### CODE REVIEW

File: framework/src/vectorpostprocessors/VolumeHistogram.C [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+  _bin_center.resize(_nbins);
+  for (unsigned i = 0; i < _nbins; ++i)
+    _bin_center[i] = (i + 0.5) * _deltaV + _min_value;
+}
+
+void
+VolumeHistogram::initialize()
+{
+  // reset the histogram
+  _volume.assign(_nbins, 0.0);
```
Comment (permcody) [inline framework/src/vectorpostprocessors/VolumeHistogram.C]: Nice!

File: framework/src/vectorpostprocessors/VolumeHistogram.C [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    // add the volume contributed by the current quadrature point
+    if (bin >= 0 && bin < _nbins)
+      _volume[bin] += computeVolume();
+  }
+}
+
+void
+VolumeHistogram::finalize()
+{
+  gatherSum(_volume);
```
Comment (permcody) [inline framework/src/vectorpostprocessors/VolumeHistogram.C]: I hate gatherSum!

File: framework/src/vectorpostprocessors/VolumeHistogram.C [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+}
+
+void
+VolumeHistogram::threadJoin(const UserObject & y)
+{
+  const VolumeHistogram & uo = static_cast<const VolumeHistogram &>(y);
+  mooseAssert(uo._volume.size() == _volume.size(), "Inconsistent volume vector lengths across threads.");
+
+  for (unsigned int i = 0; i < _volume.size(); ++i)
```
Comment (permcody) [inline framework/src/vectorpostprocessors/VolumeHistogram.C]: Not going to use `beginindex()`?

Comment (permcody) [inline framework/src/vectorpostprocessors/VolumeHistogram.C]: In fact this could just be done with a range based for loop

**Review by permcody — COMMENTED** [review]
Looks ok, You'll have to dig a little deeper to figure out why this doesn't work with threads.

File: framework/src/vectorpostprocessors/VolumeHistogram.C (hunk 2 above)
Comment (dschwen) [inline framework/src/vectorpostprocessors/VolumeHistogram.C]: but.. but.. you not only allow it, you provide it. It is a freaking MOOSE function!!!!

File: framework/include/vectorpostprocessors/VolumeHistogram.h [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+ */
+class VolumeHistogram : public ElementVectorPostprocessor
+{
+public:
+  VolumeHistogram(const InputParameters & parameters);
+
+  virtual void initialize();
+  virtual void execute();
+  virtual void finalize();
+  virtual void threadJoin(const UserObject & y);
```
Comment (dschwen) [inline framework/include/vectorpostprocessors/VolumeHistogram.h]: Ah, right, general vs. element vector postprocessors is still a bit muddy to me. Technically those are just user objects, right, so they should work as expected. I'll investigate.

File: framework/src/vectorpostprocessors/VolumeHistogram.C (hunk 2 above)
Comment (permcody) [inline framework/src/vectorpostprocessors/VolumeHistogram.C]: I'm just complaining. I plan to deprecate it moving forward.

File: framework/src/vectorpostprocessors/VolumeHistogram.C (hunk 3 above)
Comment (permcody) [inline framework/src/vectorpostprocessors/VolumeHistogram.C]: Sorry, missed that, scratch the range based suggestion. I don't think std::distance should be used for things like this.

File: framework/include/vectorpostprocessors/VolumeHistogram.h (hunk 4 above)
Comment (permcody) [inline framework/include/vectorpostprocessors/VolumeHistogram.h]: Yes, element types should have threadJoin defined and called on them.

File: framework/src/vectorpostprocessors/VolumeHistogram.C [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+}
+
+VolumeHistogram::VolumeHistogram(const InputParameters & parameters) :
+    ElementVectorPostprocessor(parameters),
+    _nbins(getParam<unsigned int>("bin_number")),
+    _min_value(getParam<Real>("min_value")),
+    _max_value(getParam<Real>("max_value")),
+    _deltaV((_max_value - _min_value) / _nbins),
+    _value(coupledValue("variable")),
+    _bin_center(declareVector(getVar("variable", 0)->name())),
```
Comment (permcody) [inline framework/src/vectorpostprocessors/VolumeHistogram.C]: It is, it's just that if you plan to couple to this vector having the name change out from underneath you will be a bit of a pain. If that's not your use case, no problem. Just a suggestion.

Comment (dschwen) [inline framework/src/vectorpostprocessors/VolumeHistogram.C]: Yeah, let's keep it as it is. The same pattern is uses in `SamplerBase` and similarly in `VectorOfPostprocessors`.

File: framework/src/vectorpostprocessors/VolumeHistogram.C (hunk 2 above)
Comment (permcody) [inline framework/src/vectorpostprocessors/VolumeHistogram.C]: To be clear, I'm not asking you to change this on this PR.

To answer your question:
`_communicator.sum()` 
I don't see any purpose to wrapping a very, very small subset of the methods available on the communicator. It doesn't make anything "easier" or clearer. I'd rather see that the communicator object is involved and just go look at the available methods. Also if you need to do a sum and then some other operation that's not wrapped in the same post processor should you mix these styles?

File: framework/include/vectorpostprocessors/VolumeHistogram.h [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+  unsigned int _qp;
+
+  /// value mid point of the bin
+  VectorPostprocessorValue & _bin_center;
+
+  /// local thread copy of the volume vector
+  VectorPostprocessorValue _volume_tmp;
+
+  /// aggregated global volume vector
+  VectorPostprocessorValue & _volume;
```
Comment (permcody) [inline framework/include/vectorpostprocessors/VolumeHistogram.h]: @friedmud - Look at the implementation here: 

This is basically the correct way to implement a `VectorPostprocessor` given the current system constraints. We have to (or probably should) create one or more `VectorPostprocessorValue` members to use for processing information during `execute()` and/or `threadJoin()`, then we have the final member that we hold a reference to from the call to `declareVector()`.

See the rest of my comments and a suggestion for an alternative design on #7809.

File: framework/src/vectorpostprocessors/VolumeHistogram.C (hunk 2 above)
Comment (friedmud) [inline framework/src/vectorpostprocessors/VolumeHistogram.C]: See my thoughts on these convenience functions here: https://github.com/idaholab/moose/issues/7820

**Review by permcody — APPROVED** [review]
This looks good, Thanks

**Pull request merged**